plugins/aggregate/aggregate.cpp

```diff
@@ -417,22 +416,25 @@ struct aggregation {
   /// Returns the aggregated batches.
   caf::expected<transform_batch> finish() {
     VAST_ASSERT(builder_);
-    // Calculate whether a column needs to be casted ahead of time.
-    auto column_needs_cast = [&] {
-      auto result = std::vector<bool>{};
-      result.resize(builder_->num_fields(), false);
-      if (!buckets_.empty())
-        for (int column = 0; column < builder_->num_fields(); ++column)
-          result[column] = !builder_->GetField(column)->type()->Equals(
-            buckets_.begin()->second[column]->type);
-      return result;
-    }();
+    auto cast_requirements
+      = std::vector<std::optional<bool>>(builder_->num_fields(), std::nullopt);
+    // Calculate whether a column needs to be casted only once.
+    auto needs_cast = [&](int column, std::shared_ptr<arrow::Scalar>& scalar) {
+      auto cast_requirement = cast_requirements[column];
+      if (!cast_requirement) {
+        if (!scalar)
+          return false;
+        cast_requirements[column] = std::optional<bool>(
+          !builder_->GetField(column)->type()->Equals(scalar->type));
+      };
+      return *cast_requirement;
+    };
     builder_->SetInitialCapacity(detail::narrow_cast<int>(buckets_.size()));
     for (auto&& bucket : std::exchange(buckets_, {})) {
       for (int column = 0; auto& scalar : bucket.second) {
-        auto* column_builder = builder_->GetField(column++);
+        auto* column_builder = builder_->GetField(column);
         VAST_ASSERT(column_builder);
-        if (column_needs_cast[column] && scalar) {
+        if (needs_cast(column++, scalar)) {
```

Comment:
You introduced a subtle new bug here: `needs_cast` must always return false if `scalar` is a `nullptr`. Now it may return `true`, which causes you to call `scalar->CastTo(...)` on a nullptr.

I'd probably change it this way:

```cpp
if (scalar && needs_cast(column, *scalar)) {
  // cast
}
++column;
```

And then change the lambda like this:

```cpp
auto needs_cast = [&](int column, const arrow::Scalar& scalar) {
  auto& cast_requirement = cast_requirements[column];
  if (!cast_requirement)
    cast_requirement =
      !builder_->GetField(column)->type()->Equals(scalar.type));
  return *cast_requirement;
};
```

Reply:
I have changed the code accordingly.